### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by atatmaja**
Sign up flow works as follows:
- Enter necessary fields (email (which is currently labelled as username but will change), password, first/last name)
- Press register
- Scan for devices
- Once a device is found, cognito sign up with initiate
- An email will be sent to your email address giving you a confirmation code, as this happens the app's ui will show an edit text where you can enter and submit your confirmation code.
- Cognito sign up will be confirmed and a post request will be made to add your user info to the userInfo table
- You will be redirected to the summary page.

Found bug:
- App crash found when creating a gson object for steps, will deal with that in a separate pr

### CODE REVIEW

**Comment by atatmaja** [comment]
Don't merge yet, still gotta change some strings and redirect to summary page after confirmation instead of going back to login

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -23,6 +23,8 @@
     <string name="continue_prompt">Continue</string>
     <string name="error_message">Authentication failure.</string>
     <string name="try_again">Incorrect username or password, please try again :)</string>
+    <string name="sign_up_error">We could not create your account, try again.</string>
```
Comment (tianyouw) [inline app/src/main/res/values/strings.xml]: Perhaps change the end of these strings to match the try_again one.

File: app/src/main/res/layout/scan_devices.xml [hunk 2]
```diff
@@ -65,4 +65,54 @@
         app:layout_constraintTop_toTopOf="parent"
         app:layout_constraintVertical_bias="0.0" />
 
+    <TextView
+        android:id="@+id/enter_code"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="8dp"
+        android:text="Enter Confirmation Code(found in email)"
```
Comment (tianyouw) [inline app/src/main/res/layout/scan_devices.xml]: Insert space before (

File: app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java [hunk 3]
```diff
@@ -650,8 +652,13 @@ public void run() {
                 //send request via apigateway
                 String response = handler.getHeartRate(userId);
                 Gson g = new Gson();
-                HeartRate hr = g.fromJson(response, HeartRate.class);
-                userHR.setCurrentHR(hr.getCurrentHR());
+                if(response != ""){
+                    HeartRate hr = g.fromJson(response, HeartRate.class);
+                    userHR.setCurrentHR(hr.getCurrentHR());
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java]: Does HR have the same GSON issue as steps?

File: app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java [hunk 4]
```diff
@@ -668,18 +675,21 @@ private void getStepDB(){
...
                 //send request via apigateway
-                handler.getStepCount(userId);
+                String response = handler.getStepCount(userId);
+                Gson g = new Gson();
+                if(response != ""){
+                    //StepCount sc = g.fromJson(response, StepCount.class);
+                    //TODO: ISSUE WITH GSON INVESTIGATE LATER
+                    userSteps.setCurrentSteps(10000);
+                }
+                else{
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java]: Change to `} else {`

File: app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java [hunk 5]
```diff
@@ -650,8 +652,13 @@ public void run() {
                 //send request via apigateway
                 String response = handler.getHeartRate(userId);
                 Gson g = new Gson();
-                HeartRate hr = g.fromJson(response, HeartRate.class);
-                userHR.setCurrentHR(hr.getCurrentHR());
+                if(response != ""){
+                    HeartRate hr = g.fromJson(response, HeartRate.class);
+                    userHR.setCurrentHR(hr.getCurrentHR());
+                }
+                else{
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java]: Change to } else {

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 6]
```diff
@@ -207,6 +325,22 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode,resultCode,data);
     }*/
 
+    //alert user of failed attempt
+    public void alertFailure(String errorMessage){
+        AlertDialog.Builder builder = new AlertDialog.Builder(this, android.R.style.Theme_Material_Light_Dialog_Alert);
+        builder.setMessage(errorMessage)
+                .setTitle(getResources().getString(R.string.error_message));
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Change to one line or multiple lines with '.'s vertically aligned.

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 7]
```diff
@@ -121,25 +213,51 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
...
+        new Thread(new Runnable(){
+            @Override
+            public void run(){
+                JSONObject userJsonObject = new JSONObject();
+                try {
+                    userJsonObject.put("userId", userId);
+                    userJsonObject.put("name", userInfo[0] + " " + userInfo[1]);
+                    //NOTE: NEED TO MODIFY FUNCTION ON BACKEND TO ENSURE THAT DEVICE ADDR GETS POSTED
+                    userJsonObject.put("deviceAddr", device_addr);
+
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Remove empty line.

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 8]
```diff
@@ -85,23 +113,87 @@ protected void onCreate(Bundle savedInstanceState) {
...
+    //TODO: MOVE COGNITO STUFF TO ITS OWN CLASS
+    // Callback handler for the sign-in process
+    AuthenticationHandler authenticationHandler = new AuthenticationHandler() {
+        @Override
+        public void onSuccess(CognitoUserSession userSession, CognitoDevice newDevice) {
+            Log.d("userSession", userSession.getUsername());
+            final String userId = userSession.getUsername();
+            CreateUserRecord(userId);
+
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Remove empty line.

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 9]
```diff
@@ -85,23 +113,87 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            }
+        }
+        @Override
+        public void onFailure(Exception exception) {
+            alertFailure(getResources().getString(R.string.sign_up_error));
+        }
+    };
+
+    GenericHandler confirmationCallback = new GenericHandler() {
+
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Empty line

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 10]
```diff
@@ -85,23 +113,87 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            if(!userConfirmed) {
+                // This user must be confirmed and a confirmation code was sent to the user
+                // cognitoUserCodeDeliveryDetails will indicate where the confirmation code was sent
+                // Get the confirmation code from user
+                newUser = cognitoUser;
+                enterCode.setVisibility(View.VISIBLE);
+                confirmationCode.setVisibility(View.VISIBLE);
+                submitButton.setVisibility(View.VISIBLE);
+            }
+            else {
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: } else {

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 11]
```diff
@@ -85,23 +113,87 @@ protected void onCreate(Bundle savedInstanceState) {
...
-                .createClient(AmazonDynamoDBClient.class,
-                        AWSMobileClient.getInstance().getCredentialsProvider(),
-                        new ClientConfiguration()
-                );
-
-        this.dynamoDBMapper = DynamoDBMapper.builder()
-                .dynamoDBClient(dynamoDBClient)
-                .awsConfiguration(AWSMobileClient.getInstance().getConfiguration())
-                .build();
+        userPool = new CognitoUserPool(getApplicationContext(), getString(R.string.cognito_userpool_id), getString(R.string.cognito_client_id), getString(R.string.cognito_client_secret), Regions.fromName(getString(R.string.cognito_region)));
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Have each param of CognitoUserPool on its own line and vertically aligned w.r.t. first argument.

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 12]
```diff
@@ -31,41 +35,65 @@
...
-    //TODO:: move
-    // Declare a DynamoDBMapper object
-    DynamoDBMapper dynamoDBMapper; //map tables to Java classes
-    AmazonDynamoDBClient dynamoDBClient;
+    private ApiGatewayHandler handler;
 
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.scan_devices);
         watch_name = getString(R.string.esperto_device_name);
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Refactor to use camel case for consistency.

File: app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java [hunk 13]
```diff
@@ -31,41 +35,65 @@
...
     boolean mServiceBound = false;
     String[] devices;
     ImageButton deviceButton;
     TextView deviceText;
+    TextView enterCode;
+    EditText confirmationCode;
+    Button submitButton;
     String[] userInfo;
     String watch_name = "Esperto";
     String device_addr;
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ScanActivity.java]: Refactor watch_name and device_addr to camel case.

File: app/src/main/java/espertolabs/esperto_ble_watch/LoginActivity.java [hunk 14]
```diff
@@ -164,12 +164,14 @@ public void run(){
...
-
-                        Intent displaySummary = new Intent(getApplicationContext(), SummaryActivity.class);
-                        displaySummary.putExtra("deviceAddress", "D4:49:8C:44:48:82");
-                        displaySummary.putExtra("user_obj", user);
-                        startActivity(displaySummary);
+                        if(user != null){
+                            Log.d("built_user", user.getName());
+
+                            Intent displaySummary = new Intent(getApplicationContext(), SummaryActivity.class);
+                            displaySummary.putExtra("deviceAddress", "D4:49:8C:44:48:82");
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/LoginActivity.java]: Should deviceAddress be a part of the user object and fetched from the server?

File: app/src/main/java/espertolabs/esperto_ble_watch/ApiGatewayHandler.java [hunk 15]
```diff
@@ -92,8 +100,33 @@ public String getUserInfo(String userId){
...
+    }
+
+    public String postUserInfo(String userJson){
+        final String invokeUrl = "https://75pp5et7e7.execute-api.us-east-1.amazonaws.com/prod/userInfo/";
+        String body = "";
+
+        try {
+            okhttp3.RequestBody req_body = okhttp3.RequestBody.create(JSON, userJson);
+            okhttp3.Request request2 = new okhttp3.Request.Builder()
+                    .url(invokeUrl)
```
Comment (tianyouw) [inline app/src/main/java/espertolabs/esperto_ble_watch/ApiGatewayHandler.java]: Align these with the '.' before Builder().

File: app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java (hunk 3 above)
Comment (atatmaja) [reply app/src/main/java/espertolabs/esperto_ble_watch/SummaryActivity.java]: Nah, it's not an issue with HR. 
The issue with steps probably due to either bad/a lack of data coming in from the request, or the fact that the StepCount class doesn't reflect the data coming in from the request.

File: app/src/main/java/espertolabs/esperto_ble_watch/LoginActivity.java (hunk 14 above)
Comment (atatmaja) [reply app/src/main/java/espertolabs/esperto_ble_watch/LoginActivity.java]: yeah it should be, thanks

**Pull request merged**